### ISSUE TITLE
Support for "not null *going forward*", for expand-contract schema+data migration patterns

### ISSUE THREAD

**Issue by aseemk** [prisma/prisma#24850]
## Problem

Prisma recommends the "expand-contract" pattern for schema & data migrations (which I agree with and think is great):

https://www.prisma.io/dataguide/types/relational/expand-and-contract-pattern

I use this pattern all the time, and the most common case of it for me is adding a new column that I want to eventually set to `NOT NULL`, but I can't at first, because a bunch of existing data still needs to be migrated to it. (And assuming that a simple backfill in a single SQL statement isn't possible, as it often isn't for non-trivial data.)

In my Prisma schema, I only have two options: fully `NOT NULL` (via no `?` modifier on the field) to require this column on both new data going forward _and existing data in the DB_ (since it adds an explicit `NOT NULL` constraint in the DB), or fully nullable (via `?` modifier), which allows both existing data to omit this column _as well as_ new data going forward.

## Suggested solution

I'd love some way to be able to specify that a field may be null on _reads_ — but must not be null on _writes_.

Specifically, no `NOT NULL` constraint should be added to the DB, but the _client library_ (generated code) can and should require the field to be explicitly provided and not null on at least creates, and maybe also updates.

Strawman examples of how this might look:

```prisma
name String? @required // to require it on all writes going forward, both creates and updates

// or:

name String? @required(creates: true, updates: false) // optional args to control when it's required
```

## Alternatives

The only alternative I'm aware of is the only option today: no automatic enforcement. Application code has to take care to ensure it's writing this column going forward on all code paths.

## Additional context

I feel like this'd be a shining example of the power of (a) generated code from Prisma's powerful custom schema language and (b) Prisma's strongly typed (TypeScript) client library! And I'd find it extremely convenient.

Thanks for your consideration.

**Comment by calum-veed** [comment]
Any update on this? I have a similar use case. 
I have an optional column which I want to be NOT NULL for all future CRUD operations. I have migrated all "active" rows and decided to leave all "legacy" rows as null. I have added a DB check constraint manually that ensures any create / update operations on rows created after a certain date have this optional row set but I'd like the Prisma client to enforce this not null constraint. Is it possible? Maybe using Prisma client extensions?

**Comment by FGoessler** [comment]
Hi,

We’ve noticed there hasn’t been activity on this issue for quite some time, and we suspect it might no longer be relevant. As part of our ongoing effort to streamline open issues in this repository, we’re closing this one to focus on what’s most important to our community.

If you believe this issue is still relevant and requires attention, feel free to reopen it. Reopening the issue is a valuable signal to us about what matters most to you.

You can learn more about our approach and motivations in the [Prisma ORM Manifesto](https://www.prisma.io/blog/prisma-orm-manifesto#1-defining-first-class-databases).

With love,
Your Prisma ORM Team ❤️

**Comment by aseemk** [comment]
Yes, this issue is still relevant and I'd still find it very convenient. It feels like it'd also play to Prisma's strength as an ORM and set it apart. Thank you.
